integration/benchmark/cmd/ci/main.go

```diff
@@ -163,57 +168,76 @@ func main() {
 	if err != nil {
 		log.Fatal().Err(err).Msgf("unable to init loader")
 	}
-	lg.Start()
-
-	testStartTime := time.Now()
-	time.Sleep(loadCase.duration)
-	lg.Stop()
-	testEndTime := time.Now()
-
-	dataSlices := calculateTpsSlices(
-		testStartTime,
-		testEndTime,
-		*leadTime,
-		*sliceSize,
-		gitSha,
-		goVersion,
-		osVersion,
-		lg)
-	prepareDataForBigQuery(dataSlices, *ciFlag)
-}
 
-func calculateTpsSlices(start, end time.Time, leadTime, sliceTime, commit, goVersion, osVersion string, lg *benchmark.ContLoadGenerator) []dataSlice {
-	//remove the lead time on both start and end, this should remove spin-up and spin-down times
-	leadDuration, _ := time.ParseDuration(leadTime)
-	endTime := end.Add(-1 * leadDuration)
-	sliceDuration, _ := time.ParseDuration(sliceTime)
+	// run load
+	lg.Start()
 
-	slices := make([]dataSlice, 0)
+	// prepare data slices
+	dataSlices := make([]dataSlice, 0)
+	defer prepareDataForBigQuery(dataSlices, *ciFlag)
 
-	for currentTime := start.Add(leadDuration); currentTime.Add(sliceDuration).Before(endTime); currentTime = currentTime.Add(sliceDuration) {
-		sliceEndTime := currentTime.Add(sliceDuration)
+	defer lg.Stop()
 
-		inputTps := lg.AvgTpsBetween(currentTime, sliceEndTime)
-		proStart := getPrometheusTransactionAtTime(currentTime)
-		proEnd := getPrometheusTransactionAtTime(sliceEndTime)
+	sliceDuration, _ := time.ParseDuration(*sliceSize)
 
-		outputTps := (proEnd - proStart) / (sliceDuration).Seconds()
+	go recordTransactionData(
```

Comment:
btw, I don't think you need  to use a goroutine here.  seems like you just want recordTransactionData to exit after some time.  you can pass a context (configured with `context.WithTimeout(...)`) into recordTransactionData, and check if the context is cancelled as part of select

this would simplify things a bit, and you don't have to worry about thread safety

Reply:
I was thinking of using a `context.WithTimeout(...)` for removing the time.Sleep() at line 169 so this aligns well with that. Will do.

Reply:
With Alexey's suggestion to swap to using channels to check for the timeout and cancelations to remove our `Sleep()` calls this no longer fits cleanly and from what I can tell we will still require the subroutine.

Reply:
You do not need a goroutine here: you can jut return `dataSlices` from this function.